sourcespec/ssp_process_traces.py

```diff
@@ -292,22 +292,42 @@ def _define_signal_and_noise_windows(config, trace):
             'Using (Ts-Tp)/2 instead')
     t1 = s_arrival_time - s_pre_time
     t1 = max(trace.stats.starttime, t1)
-    t2 = t1 + config.win_length
+    tt_s = trace.stats.travel_times['S']
+    # manage case where variable_win_length_factor is None:
+    variable_win_length_factor = config.variable_win_length_factor or 0
+    win_length_s = max(config.win_length, variable_win_length_factor * tt_s)
+    t2 = t1 + win_length_s
+    t2 = min(trace.stats.endtime, t2)
+    win_length_s = t2 - t1
     trace.stats.arrivals['S1'] = ('S1', t1)
     trace.stats.arrivals['S2'] = ('S2', t2)
     # Signal window for spectral analysis (P phase)
     t1 = p_arrival_time - config.signal_pre_time
     t1 = max(trace.stats.starttime, t1)
-    t2 = t1 + min(config.win_length, s_minus_p + s_pre_time)
+    tt_p = trace.stats.travel_times['P']
+    win_length_p = max(config.win_length, variable_win_length_factor * tt_p)
+    t2 = t1 + min(win_length_p, s_minus_p + s_pre_time)
+    t2 = min(trace.stats.endtime, t2)
+    win_length_p = t2 - t1
     trace.stats.arrivals['P1'] = ('P1', t1)
     trace.stats.arrivals['P2'] = ('P2', t2)
     # Noise window for spectral analysis
-    t1 = max(trace.stats.starttime, p_arrival_time - config.noise_pre_time)
-    t2 = t1 + config.win_length
-    if t2 >= p_arrival_time:
-        logger.warning(f'{trace.id}: noise window ends after P-wave arrival')
-        # Note: maybe we should also take into account signal_pre_time here
-        t2 = p_arrival_time
+    if config.wave_type[0] == 'P':
+        win_length = win_length_p
+    elif config.wave_type[0] == 'S':
+        win_length = win_length_s
+    if config.variable_win_length_factor:
+        logger.info(f'{trace.id}: window length {win_length:.3f} seconds')
+    # use win_length if noise_pre_time is None
+    noise_pre_time = config.noise_pre_time or win_length
+    t1 = max(
+        trace.stats.starttime,
+        p_arrival_time - noise_pre_time - config.signal_pre_time
```

Comment:
Hi Kris, coming back to the analysis of this PR (sorry for the delay).

I don't understand this extra `- config.signal_pre_time`. In fact, because of it, I get different results from the previous version, when using a fixed window length.

If I remove this extra term, then the results are identical to the previous version, when using a fixed window length.

Reply:
Claudio,
This is to ensure that the noise window ends at the start of the P window (which is equal to `p_arrival_time - config.signal_pre_time`) in the case that `config.signal_pre_time` is set to zero or None and hence defaults to the length of the signal window (but note that this is in the subsequent commit).
It may indeed give different results to before (I hope only slightly different), but in the previous version there was this note:
```
        # Note: maybe we should also take into account signal_pre_time here
```
I think the noise and signal windows are consistent this way.

Alternatively, we could leave it as before, and modify the next commit as follows:
```
if config.noise_pre_time == 0:
    noise_pre_time = win_length - signal_pre_time
```
This will give the same result when window length is variable, and should not change anything when it is fixed.

Reply:
Hi Kris, sorry again for late reaction.

Does the last commit I (force) pushed make sense to you? I tried to implement your idea (except that there should be a `+` and not a `-` in the formula 😉)

Reply:
Claudio, I'm in Germany for a meeting at the moment, so it's not so easy to look at your commit. But you are right, it should be `+`.

Reply:
Claudio, I had a look at your commit and I also tested it. Everything looks fine, so I hope it can be merged now!